File: library/src/main/java/com/thesurix/gesturerecycler/transactions/InsertTransaction.java
package com.thesurix.gesturerecycler.transactions;


import com.thesurix.gesturerecycler.GestureAdapter;
import com.thesurix.gesturerecycler.GestureViewHolder;

/**
 * @author thesurix
 */
public class InsertTransaction<T> implements AdapterTransaction {

    private final GestureAdapter<T, ? extends GestureViewHolder> mAdapter;
    private final T mItem;
    private final int mPosition;

    public InsertTransaction(final GestureAdapter<T, ? extends GestureViewHolder> adapter, final T item, final int position) {
        mAdapter = adapter;
        mItem = item;
        mPosition = position;
    }

    @Override
    public boolean perform() {
        mAdapter.getData().add(mPosition, mItem);
        mAdapter.notifyItemInserted(mPosition);
        return true;
    }

    @Override
    public boolean revert() {
        final T item = mAdapter.getData().remove(mPosition);
        final boolean success = item != null;
        if (success) {
            mAdapter.notifyItemRemoved(mPosition);
        }
        return success;
    }
}

File: library/src/main/java/com/thesurix/gesturerecycler/transactions/InsertTransaction.kt
package com.thesurix.gesturerecycler.transactions


/**
 * @author thesurix
 */
class InsertTransaction<T>(private val item: T,
                           private val position: Int,
                           private val headerEnabled: Boolean) : Transaction<T> {

    override fun perform(transactional: Transactional<T>): Boolean {
        return with(transactional.data) {
            add(position, item)
            val insertedPosition = position + if (headerEnabled) 1 else 0
            transactional.notifyInserted(insertedPosition)
            true
        }
    }

    override fun revert(transactional: Transactional<T>): Boolean {
        return with(transactional.data) {
            val item = removeAt(position)
            item?.let {
                val removedPosition = position + if (headerEnabled) 1 else 0
                transactional.notifyRemoved(removedPosition)
                true
            } ?: false
        }
    }
}
